lib/Backend/Lower.cpp

```diff
@@ -3894,6 +4031,111 @@ Lowerer::GenerateProfiledNewScObjArrayFastPath(IR::Instr *instr, Js::ArrayCallSi
+        headOpnd = GenerateArrayAlloc<Js::JavascriptNativeIntArray>(instr, lengthOpnd, arrayInfo, &isZeroed, &alignedHeadSegSizeOpnd);
+
+        GenerateMemInit(dstOpnd, Js::JavascriptNativeIntArray::GetOffsetOfArrayCallSiteIndex(), IR::IntConstOpnd::New(profileId, TyUint16, func, true), instr, isZeroed);
+        GenerateMemInit(dstOpnd, Js::JavascriptNativeIntArray::GetOffsetOfWeakFuncRef(), IR::AddrOpnd::New(weakFuncRef, IR::AddrOpndKindDynamicFunctionBodyWeakRef, m_func), instr, isZeroed);
+    }
+    else if (arrayInfo && arrayInfo->IsNativeFloatArray())
+    {
+        elementSizeOpnd = IR::IntConstOpnd::New(sizeof(Js::JavascriptNativeFloatArray::TElement), TyUint8, func);
+        missingItem = GetMissingItemOpnd(IRType::TyVar, func);
+        GenerateArrayInfoIsNativeFloatAndNotIntArrayTest(instr, arrayInfo, helperLabel);
+        lengthOpnd = generateRangeCheck();
+        Assert(Js::JavascriptNativeFloatArray::GetOffsetOfArrayFlags() + sizeof(uint16) == Js::JavascriptNativeFloatArray::GetOffsetOfArrayCallSiteIndex());
+        headOpnd = GenerateArrayAlloc<Js::JavascriptNativeFloatArray>(instr, lengthOpnd, arrayInfo, &isZeroed, &alignedHeadSegSizeOpnd);
+
+        GenerateMemInit(dstOpnd, Js::JavascriptNativeFloatArray::GetOffsetOfArrayCallSiteIndex(), IR::IntConstOpnd::New(profileId, TyUint16, func, true), instr, isZeroed);
+        GenerateMemInit(dstOpnd, Js::JavascriptNativeFloatArray::GetOffsetOfWeakFuncRef(), IR::AddrOpnd::New(weakFuncRef, IR::AddrOpndKindDynamicFunctionBodyWeakRef, m_func), instr, isZeroed);
+    }
+    else
+    {
+        elementSizeOpnd = IR::IntConstOpnd::New(sizeof(Js::JavascriptArray::TElement), TyUint8, func);
+        missingItem = GetMissingItemOpnd(IRType::TyVar, func);
+        lengthOpnd = generateRangeCheck();
+        headOpnd = GenerateArrayAlloc<Js::JavascriptArray>(instr, lengthOpnd, arrayInfo, &isZeroed, &alignedHeadSegSizeOpnd);
+    }
+    //        MOV s13,          0                         ; index
+    //        MOV s14,          16                        ; headSegmentIndex
+
+    //  ; Need to hoist missingItem outsude loop else register allocator might reuse the same register to save the hoisted missingItem
+    //  ; as that of loopIndex since missingItem is not marked live on back edge.
+    //        MOV s15,          0x80000002                ; missingItem
+
+    //$LoopTop:
+    //        CMP s13,          alignedHeadSegSizeOpnd
+    //        JGE $done
+    //        MOV [s12+s14],    0x80000002                ; Save missing element
+    //        s14               = ADD s14, 4              ; headSegmentIndex += 4. NOTE: value 4 could be different for different type of arrays
+    //        s13               = ADD s13, 1              ; index ++
+    //        JMP $LoopTop
+    //
+    //$done
+
+    IR::RegOpnd * indexOpnd = IR::RegOpnd::New(TyUint32, func);
```

Comment:
I debugged through the code for constant 8 element and found out that for size <= 8, we allocate from 3 different buckets.

constant | allocation size | bucket
----------|-----------------|---------
0,3,4,5,6 | 0x70 bytes | 6 **
1,2 | 0x60 bytes | 5
7,8 | 0x80 bytes | 7

** size 0 falls in this bucket because we allocate `SMALL_CHUNK_SIZE` as mentioned above.

We hard-code the allocation size in JIT code, but if the value is `variable`, I have to calculate the allocation size in JIT. However there are couple of things that I can do here:
1. For `0` replace `SMALL_CHUNK_SIZE` with size that is used today by array of size `1` or `2`. So 0 elements array will get allocated from bucket 5. Alternatively, I could just use `SMALL_CHUNK_SIZE` for size `1` and `2` which will make them allocate from bucket 6. In that case for array size `<= 8 ` we will be just allocate from 2 buckets. 
2. Once that is done, I can replace the allocation size and aligned size calculate code in JIT with  conditional hard-coded allocation size code. Something like this.
```asm
CMP arrayLen, 2 ; assuming 0 is moved to bucket 5
JLE $bucket1
CMP arrayLen, 6
JLE $bucket2
MOV allocSize, 0x80
MOV alignedSize, 0xA
JMP $continue

$bucket1:
   MOV allocSize, 0x60 
   MOV alignedSize, 0x2
   JMP $continue
$bucket2:
   MOV allocSize, 0x70
   MOV alignedSize, 0x6

$continue:
...
```

Let me know your thoughts.